dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp

```diff
@@ -69,14 +69,16 @@ static bool worthConvertingToLiteral(const Block & scalar)
 
 void ExecuteScalarSubqueriesMatcher::visit(const ASTSubquery & subquery, ASTPtr & ast, Data & data)
 {
-    auto hash = subquery.getTreeHash();
-    auto scalar_query_hash_str = toString(hash.first) + "_" + toString(hash.second);
+    std::stringstream out;
```

Comment:
Why tree hash is not sufficient?
Does it mean that we don't calculate it right and it should be fixed instaed?

Reply:
for example we don't have all the fields hashed in https://github.com/yandex/ClickHouse/blob/master/dbms/src/Parsers/ASTOrderByElement.h#L14

AST hashing currently only takes children into account and the only special implementation is ASTLiteral https://github.com/yandex/ClickHouse/blob/master/dbms/src/Parsers/ASTLiteral.cpp#L10

Some code refactoring needs to be done if we really want distinguishable ast hashing, and some reflection framework might be needed

Reply:
AST hash is intended to be different for different ASTs. It should be fixed.

Reflection framework will be overkill for that, because I don't know any convenient, well known and widely used C++ reflection frameworks.

Reply:
Ok, I'll fix this if there is no undergoing parser refactoring ....

Reply:
There is no active parser refactorings right now.

Reply:
Can we afford to use IAST::format to generate hashes from? Hashing through values is not enough as different types might have the same memory representation and it's too complex to maintain 60+ ASTs

Reply:
It was implemented that way and I remember that I have added separate method for hashing because it was too bad for very long querie or for queries with very long arrays / sets.